fix tests float16 module losses

MrShevan opened the pull request:
#### Changes
<!-- Please include a summary of the change and which issue is fixed. -->
<!-- Please also include relevant motivation and context. -->
<!-- List any dependencies that are required for this change. -->

Fixes #1805


#### Type of change
<!-- Please delete options that are not relevant. -->
- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update


#### Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

#### Progress

Fixed all errors when calling `pytest test/losses -v --dtype=float16 --device=cuda`, most caused by using float type in assert_close instead of torch.float.

I continue to think about using BaseTester and the mechanism that chooses the tolerance.

MrShevan commented:
Progress

Fixed all errors when calling pytest test/losses -v --dtype=float16 --device=cuda, most caused by using float type in assert_close instead of torch.float.

I continue to think about using BaseTester and the mechanism that chooses the tolerance.

MrShevan commented:
Hi @edgarriba sorry for late, I was sick(

I researched different cases of errors for float16 tensors and there are some variations:

1. Exceptions appear if arguments in `assert_close` are floats instead of tensors, so in this case selection of tolerance not possible. Most of these errors I met in `test_losses.py`, I fixed them by not casting to float:
Example:
```
-- actual = kornia.losses.kl_div_loss_2d(input, target).item()
++ actual = kornia.losses.kl_div_loss_2d(input, target)
++ expected = torch.tensor(expected).to(device, dtype)
assert_close(actual, expected)
```

2. There are fixed `atol` and `rtol` in some test cases. This also does not allow automatically to select the tolerance based on precision. So, I thought that `assert_close` should be method in `BaseTester` and don't receive as arguments `rtol` and `atol`. I implemented this mechanism of tolerance search in `BaseTester` and made all classes which uses `assert_close` as inherit from Base Tester. 

I made sure all tests pass. But some tests required lower tolerance value, such as `test_forth_and_back`. I solved it by argument `low_tolerane` in `assert_close` method, but maybe there is a better solution.

edgarriba commented on kornia/testing/__init__.py:
```diff
@@ -76,30 +76,44 @@ def create_random_fundamental_matrix(batch_size, std_val=1e-3):
+    def test_module(self, device, dtype):
         raise NotImplementedError("Implement a stupid routine.")
 
+    def assert_close(self, actual: torch.Tensor, expected: torch.Tensor, low_tolerance: bool = False) -> None:
```
```suggestion
    def assert_close(self, actual: Tensor, expected: Tensor, low_tolerance: bool = False) -> None:
```
`from kornia.core import Tensor`

edgarriba commented on test/augmentation/test_augmentation.py:
```diff
@@ -961,8 +961,33 @@ def test_random_vflip_coord_check(self, device, dtype):
         ).all()
 
+    @pytest.mark.skip(reason="not implemented yet")
+    @pytest.mark.nn
```
```suggestion
```
not very clear we need this anymore, unless we find a real utility

MrShevan replied:
Thanks for noticing, i will fix it everywhere

edgarriba commented on kornia/testing/__init__.py:
```diff
@@ -76,30 +76,44 @@ def create_random_fundamental_matrix(batch_size, std_val=1e-3):
 
     @abstractmethod
-    def test_gradcheck(self):
+    def test_gradcheck(self, device, dtype):
```
```suggestion
    def test_gradcheck(self, device):
```
I believe that for `gradcheck` we should test for `torch.float64`

edgarriba reviewed (approved):
This is an amazing improvement. I'm missing just a small thing here to somehow mark the tests we want to include in the ci. For example, to already include the testing of the functionality that already supports for half precision. Possibly imporove some logics here: https://github.com/kornia/kornia/blob/master/conftest.py

We can merge this PR and improve later

MrShevan commented:
@edgarriba I am very pleased to hear your feedback! :)
 I will work a little more on this PR (I saw some errors when you run ci) and how to improve `conftest.py`.

MrShevan commented:
@edgarriba I've done in this PR all I wanted) Now it could be merged from my side. 

I couldn't get an idea of how to ark all ready tests, but I added functionality to skip all tests, which not working for half-precision. I believe we can run CI tests by modules. I have convinced that all modules I have changed working with half-precision.

I think I can fix all the rest of modules in my following PRs :)

**Augmentation**
```
pytest test/augmentation/test_augmentation.py -v --device cuda --dtype float16
...
========= 198 passed, 86 skipped, 16 xfailed, 10 warnings in 16.89s =========
```

**Color**
```
pytest test/color/* -v --device cuda --dtype float16
...
========= 281 passed, 11 skipped in 12.98s =========
```

**Enhance**
```
pytest test/enhance/test_normalize.py  -v --device cuda --dtype float16
...
========= 33 passed, 7 skipped in 7.79s =========
```

```
pytest test/enhance/test_adjust.py  -v --device cuda --dtype float16
...
========= 128 passed, 47 skipped in 10.57s =========
```

```
pytest test/enhance/test_equalization.py  -v --device cuda --dtype float16
...
========= 24 passed, 1 skipped in 9.20s =========
```

**Losses**
```
pytest test/losses/* -v --device cuda --dtype float16
...
========= 120 passed, 1 warning in 14.77s =========
```

edgarriba commented on test/color/test_hls.py:
```diff
@@ -91,9 +91,12 @@ def test_unit(self, device, dtype):
 
     def test_nan_rgb_to_hls(self, device, dtype):
+        if dtype == torch.float16:
+            pytest.skip('not work for half-precision')
```
@MrShevan can you open issues and tag as improvements (not bugs) with the all the functions not supported by half precision?

MrShevan replied:
https://github.com/kornia/kornia/issues/1830 Done it! But I can't add label `improvement`

MrShevan commented:
I see failings in checks :( 
Looks like I have different package versions at my local ... I investigate the nature of these errors.

MrShevan commented:
@edgarriba I checked these fails and most of them related to `float64` precision. I ran only for `float12` and `float32`, so that's why I didn't notice it.

I can't decide which option better:
1. Change range for `float64` from `(1e-5, 1e-8)` to `(1e-5, 1e-4)`. I checked it solves problem and tests pass. How do you think is it enough for testing (by default `assert_close` assign (1e-7, 1e-7) to `float64`)?
2. These fails are caused by the fact that they used to be fixed `rtol` and `atol` in tests, before my changes. So I can skip these tests and create issue like I've done for `float32`.

MrShevan commented:
@edgarriba Could you approve running workflows for tests?
